File: python/raytracer.py
import array
import math
import struct

def write_bmp(fname, colors):
  h = len(colors)
  if not h: return
  w = len(colors[0])
  for r in colors:
    assert len(r) == w
  #first calculate some useful values
  rowsize = 4*(((24*w) + 31)/32)
  filesize = 54 + rowsize*h
  rowpadding = (4 - (3*w % 4)) % 4
  header_format = [
    ("Magic number", "cc"), #BM magic number
    ("File size", "i"), #size
    ("reserved 1", "H"),
    ("reserved 2", "H"),
    ("data offset", "i"),
    ("header size", "I"),
    ("width px", "I"),
    ("height px", "I"),
    ("color planes", "H"),
    ("px depth", "H"),
    ("compression method", "I"),
    ("image size", "I"),
    ("horiz res", "I"),
    ("vertical res", "I"),
    ("num colors", "I"),
    ("num important colors", "I")
  ]
  default_values = {
    'Magic number': 'BM',
    'File size': [filesize],
    'reserved 1': [0],
    'reserved 2': [0],
    'data offset': [54],
    'header size': [40],
    'width px': [w],
    'height px': [h],
    'color planes': [1],
    'px depth': [24],
    'compression method': [0],
    'image size': [0],
    'horiz res': [2835],
    'vertical res': [2835],
    'num colors': [0],
    'num important colors': [0]
  }
  print "FILE SIZE:", filesize
  print "ROW SIZE:", rowsize
  print "ROW PADDING", rowpadding
  zero = chr(0)
  buf = array.array('c', [zero for i in range(filesize)])
  offset = 0
  for k,v in header_format:
    struct.pack_into('<' + v, buf, offset, *default_values[k])
    offset += struct.calcsize(v)
  for r in colors:
    for c in r:
      for k in range(3):
	buf[offset + k] = chr(int(255.0*c.channels[2 - k]))
      offset += 3
    for k in range(rowpadding):
      buf[offset] = chr(0)
      offset += 1
  outfile = open(fname, 'wb')
  buf.tofile(outfile)
  outfile.close()   

def calc_rel_intensity2(dist2):
  return min([1.0, 3.75/(dist2)])

class Vector3(object):
  __slots__ = ['arr']
  def __init__(self, x, y, z):
    self.arr = array.array('d', [x,y,z])
    object.__init__(self)
  def setx(self,x): self.arr[0] = x
  def sety(self,y): self.arr[1] = y
  def setz(self,z): self.arr[2] = z
  def getx(self): return self.arr[0]
  def gety(self): return self.arr[1]
  def getz(self): return self.arr[2]
  def dot(self,other):
    ret = 0.0
    for a,b in zip(self.arr, other.arr):
	ret += a*b
    return ret
  def cross(self, other):
    return Vector3(self.gety()*other.getz() - other.gety()*self.getz(),
	self.getx()*other.getz() - other.getx()*self.getz(),
	self.getx()*other.gety() - other.getx()*self.gety())
  def sum(self,other):
    return Vector3(*[a+b for a,b in zip(self.arr, other.arr)])
  def difference(self,other):
    return Vector3(*[a-b for a,b in zip(self.arr, other.arr)])
  def mag2(self):
    return reduce(lambda acc,x: acc + x*x, self.arr, 0.0)
  def magnitude(self):
    return math.sqrt(self.mag2())
  def norm(self):
    divisor = self.magnitude()
    return Vector3(*[ele/divisor for ele in self.arr])
  def scale(self, scalar):
    return Vector3(*[ele*scalar for ele in self.arr])
  def equal(self, other, tolerance = 0.0001):
    for a,b in zip(self.arr, other.arr):
      if abs(a-b) > tolerance:
	return False
    return True
  def __add__(self,other):
    return self.sum(other)
  def __sub__(self,other):
    return self.difference(other)
  def __iadd__(self,other):
    for a,b,ndx in zip(self.arr,other.arr,range(3)):
      self.arr[ndx] = a+b
  def __isub__(self,other):
    for a,b,ndx in zip(self.arr,other.arr,range(3)):
      self.arr[ndx] = a-b
  x = property(getx,setx)
  y = property(gety,sety)
  z = property(getz,setz)

class Color(object):
  def __init__(self, *args):
    if len(args) != 3:
      raise ValueError, "Only 3-color schemes are supported."
    if isinstance(args[0], int):
      self.channels = self._parseIntChannels(args)
    elif isinstance(args[0], float):
      self.channels = self._parseFloatChannels(args)
    else:
     raise TypeError, "Unrecognized channel type: %s" % str(type(args[0]))      
  def _parseIntChannels(self, args):
     ret = []
     for ele in args:
      if not isinstance(ele,int):
	raise TypeError, "Cannot mix integer and float channel values."
      else:
	ret.append(max([min([1.0, float(ele)/255.0]), 0.0]))
     return ret
  def _parseFloatChannels(self, args):
     for ele in args:
      if not isinstance(ele,float):
	  raise TypeError, "Cannot mix integer and float channel values."
     constrain = lambda n: max([0.0, min([1.0,n])])
     return [constrain(n) for n in args]
  def scale(self, scalar):
    return Color(*[scalar*ele for ele in self.channels])
  def as_int_bigendian(self):
    pass
  def as_int(self):
    pass

class Shape(object):
  def __init__(self, v3, **kwargs):
    if not isinstance(v3, Vector3):
      raise TypeError, "Positional argument to Shape object must be a Vector3."
    self.pos = v3
    self.color = kwargs.pop('color')
  def intersect(self, source, dir):
      raise NotImplementedError, "Shape base class intersect method must be overridden."
  def normal(self, isect):
      raise NotImplementedError, "Shape base class normal method must be overridden."
    
class Light(object):
  def __init__(self, position, **kwargs):
    self.intensity = kwargs.pop('intensity', 1.0)
    self.position = position
    
class Camera(object):
  def __init__(self, pos, front, up, **kwargs):
    if not isinstance(front, Vector3):
      raise TypeError, "Argument 'front' must be a Vector3."
    if not isinstance(up, Vector3):
      raise TypeError, "Argument 'up' must be a Vector3."
    self.front = front.norm()
    self.up = front.cross(up).cross(front).norm() #make sure it's pendicular
    self.pos = pos #could check if it's a vector3
    self.viewing_angle_x = kwargs.pop('viewing_angle_x', math.pi/3.0)
  def get_rays(self, w, h):
    w = float(w)
    h = float(h)
    right = self.front.cross(self.up)
    viewing_angle_y = h*self.viewing_angle_x/w
    ret = []
    xincr = self.viewing_angle_x/w
    yincr = viewing_angle_y/h
    for xndx in range(int(-w/2), int(w/2)):
      for yndx in range(int(-h/2), int(h/2)):
	ret.append((
	    right.scale(math.sin(xincr*xndx*self.viewing_angle_x))
	  + self.up.scale(math.sin(yincr*yndx*viewing_angle_y)) \
	  + self.front).norm())
    return ret
  def set_front(self, fr):
     self.front = fr.norm()    

class Scene(object):
    def __init__(self, **kwargs):
	self.shapes = []
	self.lights = []
	self.camera = kwargs.pop('camera')
	self.bg_color = kwargs.pop('bg_color')
	self.ambient_lighting = 0.1
    def render(self, w, h):
      out = []
      rays = self.camera.get_rays(w,h)
      source = self.camera.pos
      tohandle = []
      count = 0
      for r in rays:
	intersections = []
	if count % w == 0:
	  out.append([])
	count += 1
	for s in self.shapes:
	  isect = s.intersect(source, r)
	  if isect is not None:
	    dist2 = (isect - source).dot(r)
	    if dist2 > 0:
	      intersections.append((dist2, isect, s))
	if len(intersections) > 0:
	  intersections.sort()
	  out[-1].append(self.get_color(intersections[0], source))
	else:
	  out[-1].append(None)
      for r in out:
	for c,cndx in zip(r, range(len(r))):
	  if c is None:
	    r[cndx] = self.bg_color
      return out
    def get_color(self, isect_eles, ray_source):
      dist, v3, shape = isect_eles
      ret = 0.0
      for l in self.lights:
	source = l.position
	direction = v3 - source
	ndir = direction.norm()
	light_dist2 = direction.mag2()
	found = False
	for s in self.shapes:
	  isect = s.intersect(source, ndir)
	  if isect is not None:
	    dist2 = (l.position - isect).mag2()
	    if dist2 - light_dist2 < -0.1:# and not v3.equal(isect):
	      found = True
	      break
	      #pass
	if not found:
	  shade = (v3 - source).norm().dot(shape.normal(v3).norm())
	  if shade < 0:	
	    shade = 0
	  ret += shade*(1.0-self.ambient_lighting) + self.ambient_lighting
	else:
	  ret += self.ambient_lighting
      return shape.color.scale(ret)
      
class Sphere(Shape):
  def __init__(self, v3, radius, **kwargs):
    self.radius = radius
    Shape.__init__(self, v3, **kwargs)
  def intersect(self, source, direction):
    square = lambda a: a*a
    assert abs(direction.mag2() - 1.0) < 0.001
    v = direction
    cntr = self.pos - source
    b = v.dot(cntr)
    discrim = square(b) - cntr.mag2() + self.radius
    if discrim < 0:
      return None
    sqdiscrim = math.sqrt(discrim)
    d1,d2 = b - discrim, b + discrim
    if d1 < d2 and d1 > 0:
      vscale = d1
    elif d2 < d1 and d2 > 0:
      vscale = d2
    else:
      return None
    return v.scale(vscale) + source
  def normal(self, isect):
    return (self.pos - isect).norm()

class Plane(Shape):
  def __init__(self, v3, norm, **kwargs):
    if not isinstance(norm, Vector3):
      raise TypeError, "Norm argument must be a Vector3 instance."
    self.norm = norm.norm()
    Shape.__init__(self, v3, **kwargs)
    self.d = self.pos.dot(self.norm)
  def intersect(self, source, direction):
    ndir = direction.norm()
    denom = ndir.dot(self.norm)
    if abs(denom) < 0.0001:
      return None
    t = (self.d - source.dot(self.norm)) / denom
    if t <= 0.0001:
      return None
    return ndir.scale(t) + source
  def normal(self, isect):
    return self.norm.norm().scale(1.0)

if __name__ == '__main__':
    origin = Vector3(0.0,0.0,0.0)
    e0 = Vector3(1.0, 0.0, 0.0)
    e1 = Vector3(0.0, 1.0, 0.0)
    e2 = Vector3(0.0, 0.0, 1.0)
    
    red = Color(1.0, 0.0, 0.0)
    green = Color(0.0, 1.0, 0.0)
    blue = Color(0.0, 0.0, 1.0)
    grey = Color(.25, .25, .25)
    white = Color(1.0, 1.0, 1.0)
    
    cam = Camera(origin, e0, e1)
    light1 = Light(e0.scale(5.0) + e2.scale(-9.0), intensity=1.0)
    sph = Sphere(e0.scale(5.0), 3.0, color=red)
    pl = Plane(e0.scale(9.0), e0.scale(1.0), color=white)
    pl2 = Plane(e2.scale(3.0), e2, color=green)
    sc = Scene(camera=cam, bg_color=green)
    sc.lights.append(light1)
    sc.shapes.append(sph)
    sc.shapes.append(pl)
    sc.shapes.append(pl2)
    write_bmp("./out_python.bmp", sc.render(512,512))
  

File: ocaml/raytracer.ml
let _COORS_LIST = [0;1;2];;
let _PI = 3.1415926535879626;;
let _DRAW_DISTANCE = 999999999999.0;;

let square x = x*.x;;
let l2distance source dest =
   List.fold_left (fun a b -> a +. b) 0.0 (List.map square (List.map (fun n -> (source#getcoor n) -. (dest#getcoor n)) _COORS_LIST))
;;
let norm_scalar iger minval maxval =
  (float_of_int iger) /. (float_of_int (maxval - minval))
;;

class color r g b =
  object(self)
    val mutable my_r = r land 0xFF
    val mutable my_g = g land 0xFF
    val mutable my_b = b land 0xFF
    method as_int =
      (my_r lsl 16) + (my_g lsl 8) + my_r
    method getr = my_r
    method getg = my_g
    method getb = my_b
  end;;

let blend c1 c2 =
  new color (c1#getr + c2#getr) (c1#getg + c2#getg) (c1#getb + c2#getb)
;;
(*let project lgt surface =
  let tot = float_of_int (surface#getr + surface#getg + surface#getb) in
  new color (int_of_float ((float_of_int lgt#getr) *. (float_of_int surface#getr)/.tot))
      (int_of_float ((float_of_int lgt#getg) *. (float_of_int surface#getg)/.tot))
      (int_of_float ((float_of_int lgt#getb) *. (float_of_int surface#getb)/.tot))
;;*)
let project lgt surface =
  lgt
;;
(*
let dim c factor =
  let recip = int_of_float (1.0/.factor) in
  new color (c#getr / recip) (c#getg / recip) (c#getb / recip)
;;
*)
let dim c factor =
  let convert channel =
    int_of_float ((float_of_int channel) *. factor)
  in
  new color (convert c#getr) (convert c#getg) (convert c#getb)
;;

class vector3 =
  object(self)
      val mutable my_x = 0.0
      val mutable my_y = 0.0
      val mutable my_z = 0.0
    method set_position xin yin zin =
      my_x <- xin;
      my_y <- yin;
      my_z <- zin;
      self
    method set_v3_position (v3 : vector3) =
      self#set_position (v3#getx) (v3#gety) (v3#getz)
    method getcoor n =
      match n with
    	| 0 -> my_x
	    | 1 -> my_y
	    | 2 -> my_z
	    | _ -> raise Not_found
    method setcoor n coor =
      match n with
	    | 0 -> my_x <- coor; ()
	    | 1 -> my_y <- coor; ()
	    | 2 -> my_z <- coor; ()
	    | _ -> raise Not_found	
    method list_coors =
      List.map (fun x -> self#getcoor x) _COORS_LIST
    method sum (v3 : vector3) =
      let ret = new vector3 self#getcolor in
      List.map (fun n -> ret#setcoor n ((self#getcoor n) +. (v3#getcoor n))) _COORS_LIST;
      ret
    method inverse =
      let ret = new vector3 self#getcolor in
      List.map (fun n -> ret#setcoor n (0.0 -. (self#getcoor n))) _COORS_LIST;
      ret
    method difference (v3: vector3) =
      self#sum (v3#inverse)
    method magnitude =
      sqrt (List.fold_left (fun acc x -> acc +. x*.x) 0.0 (List.map (fun n -> self#getcoor n) _COORS_LIST))
    method normalize =
      let ret = new vector3 self#getcolor in
      let mag = self#magnitude in
      List.map (fun n -> ret#setcoor n ((self#getcoor n) /. mag)) _COORS_LIST;
      ret
    method dot (v3: vector3) =
      List.fold_left2 (fun acc l r -> acc +. (l*.r)) 0.0 (v3#list_coors) (self#list_coors)
    method scale x =
      let ret = new vector3 self#getcolor in
      List.map2 (fun n m -> ret#setcoor n (x*.m)) _COORS_LIST self#list_coors;
      ret
    method cross (v3: vector3) =
      let ret = new vector3 self#getcolor in
      ret#set_position ((self#gety *. v3#getz) -. (self#getz *. v3#gety)) ((self#getz *. v3#getx) -. (self#getx *. 
          v3#getz)) ((self#getx *. v3#gety) -. (self#gety *. v3#getx))
    method getx = my_x
    method gety = my_y
    method getz = my_z
    method print =
      print_endline ((string_of_float my_x) ^ ", " ^ (string_of_float my_y) ^ ", " ^ (string_of_float my_z))
  end;;

let _ZERO_VECTOR = new vector3;;

let rec extractvalid ret lst =
  match lst with
    | [] -> ret
    | _ -> (
      let curr = List.hd lst in
      match curr with
	| None -> extractvalid ret (List.tl lst)
	| Some (v3) -> extractvalid (v3 :: ret) (List.tl lst))
;;

class light (clr : color) (v3 : vector3) intensity =
  object(self)
    val mutable my_clr = clr
    val mutable my_pos = v3
    val mutable my_inty = if intensity > 1.0 then 1.0 else if intensity < 0.0 then 0.0 else intensity
    method getpos = my_pos
    method getcolor = my_clr
    method getinty = my_inty
  end;;

class intersection (pos : v3) (clr : color) (norm : v3) =
  object(self)
    val mutable my_pos = pos
    val mutable my_clr = clr
    val mutable my_norm = norm
    method getpos = my_pos
    method getcolor = my_clr
    method getnorm = my_norm
  end;;

class virtual shape x y z =
  object(self)
    val mutable center = (new vector3)#set_position x y z
    val mutable my_color = new color 255 255 255
    method getcolor = my_color
    method setcolor c = my_color = c
    method virtual intersect: vector3 -> vector3 -> (vector3 option)
  end;;

class sphere x y z radius =
  object(self)
    val mutable my_radius = radius
    inherit shape x y z
    method norm location =
      (* we assume the location actually intersects the sphere *)
      (center#difference location)#normalize
    method intersect source direction =
      let cntr = (center#difference source) in
      let dir = direction#normalize in
      let b = cntr#dot dir in
      let discriminant = (square b) -. (cntr#dot cntr) +. (radius*.radius) in
      if discriminant < 0.0 then None
      else (
	      let addend = sqrt (discriminant) in
	      let one,two = ((-.b +. addend), (-.b -. addend)) in
	      let makeret scalar =
	        let retv3 = (new vector3)#set_v3_position ((dir#scale (-.scalar))#sum source) in
	        Some (new intersection retv3 self#getcolor (self#norm retv3)) 
	      in
	      let ret =
	        if one < 0.0 && two < 0.0 then None
	        else if one > two then makeret one
	        else makeret two
	      in
	      ret)
  end;;

class shapecollection x y z =
  object(self)
    inherit shape x y z
    val mutable my_shapes = ([] : shape list)
    method intersect source direction =
      let rec helper shapes ret =
	      match shapes with
	      | [] -> ret
	      | _ -> helper (List.tl shapes) (((List.hd shapes)#intersect source direction) :: ret)
      in	  
      let distpairs = List.map (fun x -> (l2distance center x, x)) (extractvalid [] (helper my_shapes [])) in
      match distpairs with
	    | [] -> None
	    | _ -> (
	      let retv3 = (new vector3)#set_v3_position (snd (List.fold_left (fun a b -> if (fst a) < (fst b) then a else b)
	        Some ( (0.0, _ZERO_VECTOR) distpairs)))
  end;;

class camera w h =
  object(self)
    val mutable my_width = w
    val mutable my_height = h
    val mutable my_pos = _ZERO_VECTOR
    val mutable my_front = (new vector3 None)#set_position 1.0 0.0 0.0
    val mutable my_up = (new vector3 None)#set_position 0.0 1.0 0.0
    method getpos = my_pos
    method set_v3_position v3 =
      my_pos <- v3;
      self
    method getrays =
      my_front <- my_front#normalize;
      my_up <- my_up#normalize;
      let viewing_angle_x = _PI /. 2.0 in
      let viewing_angle_y = viewing_angle_x *. float_of_int(h) /. float_of_int(w) in
      let my_right = (my_front#cross my_up)#normalize in
      (*print_endline "RIGHT:";
      my_right#print;*)
      let ray xin yin =
	let x = (norm_scalar xin 0 w) -. 0.5 in
	let y = (norm_scalar yin 0 h) -. 0.5 in
	((my_front#scale (-.1.0))#sum ((my_right#scale (sin (x*.viewing_angle_x)))#sum (my_up#scale (sin (y*.viewing_angle_y)))))#normalize
      in
      let rec helper ret xrem yrem =
	match yrem with
	  | -1 -> ret
	  | _ -> (
	  match xrem with
	    | -1 -> helper ret (w-1) (yrem-1)
	    | _ -> helper ((ray xrem yrem) :: ret) (xrem-1) yrem)
      in
      helper [] (w-1) (h-1)
  end;;

class scene (bgclr : color)  (cam : camera)=
  object(self)
    val mutable my_shapes = ([] : shape list)
    val mutable my_lights = ([] : light list)
    val mutable my_bgclr = bgclr
    val mutable my_camera = cam
    method add_shape s =
      my_shapes <- s :: my_shapes
    method add_light l =
      my_lights <- l :: my_lights
    method get_lighting v3 norm =
      let rec helper rem ret =
	      match rem with
	      | [] -> ret
	      | _ ->
	        let currlight = List.hd rem in
	        let dir = (currlight#getpos)#difference v3 in
	        let isect = self#intersect_ray ((v3#scale (1.0))#sum (dir#scale (0.1))) (dir#scale (-.1.0)) in
	        match isect with
	        | None -> (
	        (*print_endline ("magnitude:" ^ (string_of_float dir#magnitude));
	        print_endline "LGT:";
	        currlight#getpos#print;25
	        print_endline "V3:";
	        v3#print;*)
	          helper (List.tl rem) (blend ret (dim currlight#getcolor (4.0 /. (square dir#magnitude)))))
	        | Some(v3_2) -> (
	          (*print_endline "GOT INTERSECT:";
	          v3_2#print;
	          v3#print;*)
	        helper (List.tl rem) ret)
        in
      project (helper my_lights (new color 25 25 25)) v3#getcolor_or_err
    method intersect_ray source direction =
      let rec find_intersections ret rem =
        match rem with
	      | [] -> ret
	      | _ -> find_intersections (((List.hd rem)#intersect source direction) :: ret) (List.tl rem)
      in
      let cmp best isect =
	      let mag = (source#difference isect)#magnitude in
	      if mag < (snd best) then (isect,mag)
	      else best
      in
      let getcmpfirst valid_list = 
	      let first = List.hd valid_list in
	      (first, (source#difference first)#magnitude)
      in
      let valid = extractvalid [] (find_intersections [] my_shapes) in
      match valid with
	    | [] -> None
	    | _ -> (
	      (*(List.nth valid 0)#print;*)
	      Some (fst (List.fold_left cmp (getcmpfirst valid) valid)))
    method handle_ray source direction =
      match self#intersect_ray source direction with
	    | None -> my_bgclr
	    | Some(v3) -> self#get_lighting v3
    
    method render =
      let rays = my_camera#getrays in
      let rec getmax rem best =
      	match rem with
	      | [] -> best
	      | _ -> (
	        let curr = List.hd rem in
	        getmax (List.tl rem) (min (curr#gety) best))
      in
      let rec helper ret rem =
	      match rem with
	      | [] -> ret
	      | _ -> (
	        let curr = (List.hd rem) in
	        helper ((self#handle_ray my_camera#getpos curr) :: ret) (List.tl rem))
        in
      helper [] rays
  end;;

let swap i =
  ((i land 0xFF) lsl 24) + (((i lsr 8) land 0xFF) lsl 16) + (((i lsr 16) land 0xFF) lsl 8) + ((i lsr 24) land 0xFF)
;;
let write_bitmap fname w h clrs =
  let f = open_out_bin fname in
  let ob b = output_byte f b in
  let oi i =
    let rec helper rem =
      match rem with
      | 0 -> ()
      | _ -> (ob ((i lsr ((4 - rem)*8)) land 0xFF);
	helper (rem - 1))
    in
    helper 4
  in
  let rowpadding = (4 - ((3*w + 4)/4 - 1)) mod 4 in
  let rec pad n =
    match n with
    | 0 -> ()
    | _ -> (
      ob 0;
      pad (n-1))
  in
  let rec writerow rem clrs =
    match rem with
    | 0 -> (pad rowpadding; clrs)
    | _ -> (
      let curr = List.hd clrs in
      ob curr#getb; ob curr#getg; ob curr#getr;
      writerow (rem - 1) (List.tl clrs))
  in
  ob 0x42; ob 0x4D; (*magic number - 2 bytes*)
  oi (54 + (4 * (24*w + 31)/32)); (* BMP file size - 4 bytes*)
  oi 0x0; (* 4 empty bytes, for application specific purposes *)
  oi 54; (*location of start of bitmap data*)
  (*START V3 HEADER *)
  oi 40; (*header size*)
  oi w; (*bitmap width in pixels*)
  oi h; (*bitmap height in pixels *)
  ob 1; ob 0; (*little-endian way to indicate 1 color plane using 2 bytes*)
  ob 24; ob 0; (*24 bit color depth. little-endian*)
  oi 0; (*no compression*)
  oi 0; (*would be the image size field if we were using compression*)
  oi 2835; (*2835 pixels per meter horiz. res.*)
  oi 2835; (*2835 pixels per meter vert. res. *)
  oi 0; (*default number of colors in palette (2^(color depth)) *)
  oi 0; (* "import color" pretty much ignored *)

  (*DONE WITH THE BITMAP HEADER*)
  let rec helper rem clrs =
    match rem with
    | 0 -> ()
    | _ -> (
      helper (rem - 1) (writerow w clrs))
  in
  helper h clrs;
  close_out f
;;
let cam = new camera 1024 640;;
let sph = new sphere 10.0 0.0 0.0 3.0;;
let sc = new scene (new color 0 255 0) cam;;
let lt = new light (new color 255 0 0) ((new vector3 None)#set_position 18.0 1.25 0.0) 1.0;;
let lt2 = new light (new color 0 0 255) ((new vector3 None)#set_position 12.0 (-.5.0) 2.0) 1.0;;
let lt3 = new light (new color 0 255 0) ((new vector3 None)#set_position 12.0 (0.0) 5.0) 1.0;;
let lt4 = new light (new color 255 255 255) ((new vector3 None)#set_position 12.0 0.0 0.0) 1.0;;
sph#setcolor (new color 255 255 255);;
sc#add_light lt;;
sc#add_light lt2;;
sc#add_light lt3;;
sc#add_light lt4;;
sc#add_shape sph;;
let img = sc#render;;
write_bitmap "./out.bmp" 1024 640 img;;
print_endline (string_of_int (List.length img))
